### DISCUSSION TITLE
Booting Discovery Kit from SD card and applying device tree overlays

### DISCUSSION THREAD

**Discussion by NavadeepGaneshU** [polarfire-soc/polarfire-soc-documentation#461] [Linux build systems for PolarFire SoC] [answered]
Hi, 
I am trying to boot the latest released image from (https://github.com/polarfire-soc/meta-polarfire-soc-yocto-bsp/releases/tag/v2024.02.1) on the MPFS Disco Kit and after inserting the SD card, nothing is showing up on the serial terminal. Is there any additional consideration for this kit?
Some points:
1. Flashed the core-image-minimal-dev-mpfs-disco-kit-20240426091646.rootfs.wic file using windows disk imager and it showed 3 sections inside the SD card.
2. Hot plugged the SD card while powered ON and connected to UART. Tried the other way as well my connecting after inserting the SD card.
3. I had programmed the (https://github.com/polarfire-soc/polarfire-soc-bare-metal-examples/blob/main/driver-examples/mss/mss-gpio/mpfs-gpio-interrupt) example before to the board and it this was working. Now when I connect the serial terminal, the serial output of this example with ' Microchip MPFS-DISCO-KIT Test' is getting printed in COM9 (among COM8/9/10). I suppose it is from eNVM and should this be cleared for linux to boot?
4. I have tried loading my own yocto build run on WSLv2 and the reference eample image given at (https://github.com/polarfire-soc/polarfire-soc-discovery-kit-reference-design/releases/tag/2024.04) as well from any of which I couldn't find the UART terminal showing up with the U-Boot loading.

**Comment by hughbreslin** [comment]
Hi @NavadeepGaneshU it sounds like you programmed the GPIO example into the eNVM? This has probably overwritten the HSS which is used as a boot loader for Linux. Can you try reprogramming the discovery kit reference design using FPExpress (this includes the HSS) and see if that helps your issue.

**Comment by NavadeepGaneshU** [comment]
Thanks for the quick response @hughbreslin. Yes, I was doing that with the GPIO example and after programming the .job file from the reference design example, it worked out. 
Next, trying to control a GPIO pin (GPIO2, PIN17) and I am having gpio@20120000, gpio@20121000, gpio@20122000 identified inside the `device-tree/soc` . Can you point me to a reasonable reference or guide how to configure this pin as GPIO and command it to write 1 and 0?
(I see mpfs-icicle-kit comes loaded with gpiotest example)

**Comment by hughbreslin** [comment, marked as answer]
No problem :) Ok cool I'm glad thats working! In our current device tree we haven't enabled the GPIOs yet (it will be done in a future release) but they are connected in the reference design. The easiest way to enable and control them is using a device tree overlay.

I have attached a sample to do this: [gpio_overlay_src.zip](https://github.com/user-attachments/files/15729514/gpio_overlay_src.zip). You'll need to extract the zip file, GitHub doesn't let me share the source directly. You should get 'gpio_overlay_src.dtso'

I used SCP to transfer this to my kit over ethernet. If you do this you can use the following steps to apply it:

1. Generate the overlay using `dtc gpio_overlay_src.dtso -o gpio_overlay_src.dtbo` - there might be some warnings, thats ok
2. Create a directory for the overlay `mkdir /sys/kernel/config/device-tree/overlays/gpio`
3. Finally apply the overlay using `cat ./gpio_overlay_src.dtbo > /sys/kernel/config/device-tree/overlays/gpio/dtbo`

This then enables the GPIO. To control it I use libgpiod. For example:

`gpioinfo`: will print out GPIO configuration
`gpioset gpiochip1 17=1`: will set what is GPIO17
gpioset gpiochip1 17=0`: will un-set what is GPIO17

GPIO17 is LED1 which you can see in the overlay file connected to MSS GPIO2 :)

I hope this helps. In the future release you will only need the `gpioinfo` and `gpioset` commands once the GPIO connections are added to the base device tree.

**Comment by NavadeepGaneshU** [comment]
Thanks for the detailed explanation. "Ask Hugh" is a pretty great phrase. Now I can use python script to toggle and control them in any order. Adding onto this, I tried extending the device-tree overlay for I2C. Knowing there is libi2cd present and I am seeing `i2c@2010a000` , `i2c@2010b000` etc, I included

```
i2c0: i2c@2010a000 {
            compatible = "microchip,mpfs-i2c";
            reg = <0x00 0x2010a000 0x00 0x1000>;
            interrupt-parent = <&irqmux>;
            interrupt-controller;
            clock-frequency = <400000>;
            i2c-controller;
            #address-cells = <1>;
            #size-cells = <0>;
            status = "okay";

            htu21d@40 {
                compatible = "te,htu21d";
                reg = <0x40>;
                status = "okay";
            };
        };
```

but the device i2c peripheral doesn't seem to appear under `ls /dev/`. Sorry I know I am going beyond what is asked in the original question, but a clarification on this will be of help in kit review and development I am doing.

**Comment by hughbreslin** [comment]
Haha great 👍 if you want you can add a script to automatically apply the overlay at boot so you don’t have to do it manually as well on each startup :)

Looking at what you shared for I2C, MSS I2C_0 is enabled on the kit. I had a look at our base device tree entry [here](https://github.com/linux4microchip/linux/blob/99b20cea90ff1a05163da777cb95b0dae2016ef1/arch/riscv/boot/dts/microchip/mpfs.dtsi#L381) and I can see a few differences: 

`
		i2c0: i2c@2010a000 {
			compatible = "microchip,mpfs-i2c", "microchip,corei2c-rtl-v7";
			reg = <0x0 0x2010a000 0x0 0x1000>;
			#address-cells = <1>;
			#size-cells = <0>;
			interrupt-parent = <&plic>;
			interrupts = <58>;
			clocks = <&clkcfg CLK_I2C0>;
			clock-frequency = <100000>;
			status = "disabled";
		};
`

I’ll be able to try it out tomorrow but I’d say get most of it lined up (and marked enabled!) and fingers crossed it should work :)

**Comment by hughbreslin** [comment]
Hey @NavadeepGaneshU 

I tried out the following:

```
			i2c0: i2c@2010a000 { 
				compatible = "microchip,mpfs-i2c", "microchip,corei2c-rtl-v7";
				reg = <0x0 0x2010a000 0x0 0x1000>;
				#address-cells = <1>;
				#size-cells = <0>;
				interrupt-parent = <&plic>;
				interrupts = <58>;
				clocks = <0x01 0xF>;
				clock-frequency = <100000>;
				status = "okay";
			};
```

And I can see I2C-0 show up in /dev/i2c* :) I don't have anything to test it with but hopefully this is enough to get you on the right track. I got the clock value from the [mpfs-clock.h file](https://github.com/linux4microchip/linux/blob/99b20cea90ff1a05163da777cb95b0dae2016ef1/include/dt-bindings/clock/microchip%2Cmpfs-clock.h#L26C18-L26C20).

This thread has actually been quite useful and has gone beyond the scope a bit but could be a useful discussion item so its more public. I'll get it converted over :)

**Comment by NavadeepGaneshU** [comment]
Thanks for following up @hughbreslin. The device gets added to the `/dev/` list now as you've directed while the reference design image is loaded. I was on the custom yocto build image made using the instructions in this repo. Having trouble in getting the correct I2C sensor address detected by the `i2detect 0` search but it can be on my side.

**Comment by zmcginty-thinkOrbital** [comment]
I'm also having the same problem of not getting any console output from my mpfs-disco-kit on any of the three serial ports, except for getting a prompt to start self-tests(DDR, SD-Card, and Ethernet). So if I'm not getting any bootloader output on any console, I need to re-flash the reference design? And to re-flash the reference design I need to download this repo; https://github.com/polarfire-soc/polarfire-soc-discovery-kit-reference-design and import it into libero (which won't open due to no license file, even though I downloaded the free version) and use Libero to flash the board?

**Comment by rockoguido** [comment]
Hi..

In order to get the Discovery Kit to a known state you can just start with a pre-built image for this kit at..

https://github.com/polarfire-soc/polarfire-soc-discovery-kit-reference-design/releases

The first one on the list is most likely sufficient for your use case

You can deploy this without the use of Libero.   You can just use FP Express, which is part of the Libero SDK

This image will setup the SoC and the Fabric and contains the stage 0 bootloader we provide, HSS.  At this point you will see HSS out of one of the 3 UARTS that is available with the Discovery Kit

Regards,

Dave Campanella
Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 989 0569 (mobile)
+1 760 304 0172 (office)
***@***.******@***.***>

For any Technical related concerns, kindly you can create web ticket at http://www.microchip.com/support
For Technical Training please go to Microchip University at https://secure.microchip.com/mu

There are several paths to take for Microchip support:
        * Self-paced learning – see the Developer Help Site<http://microchipdeveloper.com/>
        * Knowledge Base - here<http://www.microchip.com/support/hottopics.aspx>
        * Community Forums - help<http://www.microchip.com/forums/> with a lot of information from experts.

**Comment by zmcginty-thinkOrbital** [comment]
rockoguido that worked great! I re-flashed my two mpfs boards and got console output from them for first stage bootloader, and got u-boot and kernel console output on the other serial port. I am now trying to test this board with an SPI DAC, and am wondering how to access the SPI peripheral from within the linux image (default-config built with yocto). I see there's an i2c-0 in /dev, but don't see any spi devices in /dev. I'm guessing I'm missing something simple here, but it looks like there are entries for spi when I run dtc -I fs /proc/device-tree such as those below. But I don't see any spi device show up in /dev

		spi@20109000 {
			#address-cells = <0x01>;
			interrupts = <0x37>;
			clocks = <0x01 0x0e>;
			#size-cells = <0x00>;
			interrupt-parent = <0x09>;
			compatible = "microchip,mpfs-spi";
			status = "okay";
			reg = <0x00 0x20109000 0x00 0x1000>;
			phandle = <0x28>;
		};
		spi@20108000 {
			#address-cells = <0x01>;
			interrupts = <0x36>;
			clocks = <0x01 0x0d>;
			#size-cells = <0x00>;
			interrupt-parent = <0x09>;
			compatible = "microchip,mpfs-spi";
			status = "okay";
			reg = <0x00 0x20108000 0x00 0x1000>;
			phandle = <0x27>;
		};
		spi@37020100 {
			#address-cells = <0x01>;
			interrupts = <0x6e>;
			clocks = <0x12>;
			#size-cells = <0x00>;
			interrupt-parent = <0x09>;
			compatible = "microchip,mpfs-qspi\0microchip,coreqspi-rtl-v2";
			status = "disabled";
			reg = <0x00 0x37020100 0x00 0x100>;
			phandle = <0x34>;
		};
		spi@21000000 {
			#address-cells = <0x01>;
			interrupts = <0x55>;
			clocks = <0x01 0x16>;
			#size-cells = <0x00>;
			interrupt-parent = <0x09>;
			compatible = "microchip,mpfs-qspi\0microchip,coreqspi-rtl-v2";
			status = "disabled";
			reg = <0x00 0x21000000 0x00 0x1000>;
			phandle = <0x29>;
		};

**Comment by rockoguido** [comment]
Ok!

Great to see you are up and running, we live for small victories!

So, in regards to SPI and the Disco Kit you  need some device on the Mikroe connector for it to connect to.

We like here this option..
https://www.mikroe.com/sqi-flash-click?srsltid=AfmBOoqMchd-IMhwt9ZWAONOCJwHPoN4jcmN404zkPrJbBBroZ-BSsm1

Inexpensive and it has a MCHIP QSPI flash part, keeping it in the family.

As you pointed out below the device tree has SPI enabled and with this device I list, or similar, should see the device instantiated in /dev.

I hope that helps

Regards,

Dave Campanella
Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 989 0569 (mobile)
+1 760 304 0172 (office)
***@***.******@***.***>

For any Technical related concerns, kindly you can create web ticket at http://www.microchip.com/support
For Technical Training please go to Microchip University at https://secure.microchip.com/mu

There are several paths to take for Microchip support:
        * Self-paced learning - see the Developer Help Site<http://microchipdeveloper.com/>
        * Knowledge Base - here<http://www.microchip.com/support/hottopics.aspx>
        * Community Forums - help<http://www.microchip.com/forums/> with a lot of information from experts.

**Reply by zmcginty-thinkOrbital** [reply]
My problem is that no SPI devices are showing up in /dev/ or in /sys/bus/spi/devices (shown below);
root@mpfs-disco-kit:~# ls /sys/bus/spi/devices/
root@mpfs-disco-kit:~# ls /dev/spi*
ls: cannot access '/dev/spi*': No such file or directory

while i2c seems to show up in both places;
root@mpfs-disco-kit:~# ls /sys/bus/i2c/devices/
i2c-0
root@mpfs-disco-kit:~# ls /dev/i2c*  
/dev/i2c-0

**Reply by zmcginty-thinkOrbital** [reply]
I am able to see two SPI devices in /sys/class/spi_master;
root@mpfs-disco-kit:~# ls /sys/class/spi_master/
spi0  spi1

**Comment by rockoguido** [comment]
Ok what you see is as expected.

Nice to hear


Dave Campanella
Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 989 0569 (mobile)
+1 760 304 0172 (office)
***@***.******@***.***>

For any Technical related concerns, kindly you can create web ticket at http://www.microchip.com/support
For Technical Training please go to Microchip University at https://secure.microchip.com/mu

There are several paths to take for Microchip support:
        * Self-paced learning – see the Developer Help Site<http://microchipdeveloper.com/>
        * Knowledge Base - here<http://www.microchip.com/support/hottopics.aspx>
        * Community Forums - help<http://www.microchip.com/forums/> with a lot of information from experts.

**Comment by zmcginty-thinkOrbital** [comment]
So I'm guessing that since no spi devices are showing up in /dev, that I need to modify the device tree. it appears that the spi devices are enabled but from the research I've done so far, it looks like I might need to use "spidev" for compatibility, to get it to show up in /dev. Does anyone have any experience on where to create layers/add patches to the yocto repo (https://github.com/polarfire-soc/meta-polarfire-soc-yocto-bsp) in order to modify the device tree?

**Comment by rockoguido** [comment]
Hi..

To modify the device tree the layer would be  bsp layer..  meta-polarfire-soc-bsp and the recipe would be the kernel recipe.   mpfs-linux.bb
I suspect you are aware that you don’t just modify the recipe file directly, changes made via a patch file that you create, patch file is referenced in the recipe file.

Regards,

Dave Campanella
Principal Embedded Solutions Engineer / FAE
Microchip Technology Inc.
165 Technology Dr
Irvine, CA 92618
+1 925 989 0569 (mobile)
+1 760 304 0172 (office)
***@***.******@***.***>

For any Technical related concerns, kindly you can create web ticket at http://www.microchip.com/support
For Technical Training please go to Microchip University at https://secure.microchip.com/mu

There are several paths to take for Microchip support:
        * Self-paced learning – see the Developer Help Site<http://microchipdeveloper.com/>
        * Knowledge Base - here<http://www.microchip.com/support/hottopics.aspx>
        * Community Forums - help<http://www.microchip.com/forums/> with a lot of information from experts.

**Reply by zmcginty-thinkOrbital** [reply]
Rockoguido, Yes, I'm not trying to modify the device trees in place. At the moment I'm trying to create a device tree overlay and apply it at runtime using the dt-overlay script in /opt/microchip/dt-overlay. According to dmesg, it appears to be applying it, as it gives me a warning about memory overflow. However I still see no spi or spidev devices showing up in /dev/. Do you happen to know if I need to change pinmuxing, and if so, where do I change this?
